src/commonMain/kotlin/fr/acinq/eclair/db/PaymentsDb.kt

```diff
@@ -182,17 +201,44 @@ data class OutgoingPayment(val id: UUID, val recipientAmount: MilliSatoshi, val
         /** Swaps out send a lightning payment to a swap server, which will send an on-chain transaction to a given address. */
         data class SwapOut(val address: String, override val paymentHash: ByteVector32) : Details()
 
+        /** Corresponds to the on-chain payments made when closing a channel. */
+        data class ChannelClosing(
+            val channelId: ByteVector32,
+            val closingAddress: String,
+            val isSentToMyWallet: Boolean,
+            override val paymentHash: ByteVector32
+        ) : Details()
+
         fun matchesFilters(filters: Set<PaymentTypeFilter>): Boolean = when (this) {
             is Normal -> filters.isEmpty() || filters.contains(PaymentTypeFilter.Normal)
             is KeySend -> filters.isEmpty() || filters.contains(PaymentTypeFilter.KeySend)
             is SwapOut -> filters.isEmpty() || filters.contains(PaymentTypeFilter.SwapOut)
+            is ChannelClosing -> filters.isEmpty() || filters.contains(PaymentTypeFilter.ChannelClosing)
         }
     }
 
     sealed class Status {
         object Pending : Status()
-        data class Succeeded(val preimage: ByteVector32, val completedAt: Long = currentTimestampMillis()) : Status()
-        data class Failed(val reason: FinalFailure, val completedAt: Long = currentTimestampMillis()) : Status()
+        sealed class Completed : Status() {
+            abstract val completedAt: Long
+            data class Failed(val reason: FinalFailure, override val completedAt: Long = currentTimestampMillis()) : Completed()
+            sealed class Succeeded : Completed() {
+                data class OffChain(
+                    val preimage: ByteVector32,
+                    override val completedAt: Long = currentTimestampMillis()
+                ) : Succeeded()
+                data class OnChain(
+                    val txids: List<ByteVector32>,
+                    val claimed: Satoshi,
```

Comment:
You should document this field.
It looks like for this `OnChain` case, the fee we could display to the user would actually be `recipientAmount - claimed`, is that correct?

I still think the type architecture is incorrect here though. I think it should be:

```kotlin
data class OnChain(val type: ChannelClosingType, override val completedAt: Long = currentTimestampMillis()) : Succeeded()
```

And each claimed tx should be a `Part`.
It does require a bit more work (to make `Part` more abstract), but I'm afraid we'll have to pay that toll later anyway *and* a data migration...

Reply:
Actually the data migration shouldn't be too bad, so feel free to ignore this comment for now for the sake of shipping sooner ;)

Reply:
Done in 551cf9f

> I think it should be: each claimed tx should be a `Part`.

Agree 100%

> we'll have to pay that toll later anyway and a data migration...

I believe the data migration will actually be easy.

`Part` will probably become a sealed class:
```kotlin
sealed class Part {
    data class LightningPart(/* ... */) : Part()
    data class ClosingTransaction(/* ... */): Part()
}
```

But on-disk, the `ClosingTransaction` it will have it's own separate database table.

The structure of `OnChain` will change. But luckily we've already planned for this, and the database architecture is now designed to handle these upgrades:

```kotlin
// The database table has these columns:
// - status_type INTEGER DEFAULT 0
// - status BLOB DEFAULT NULL
//
// The `status_type` column is populated with a value from this enum.
//
enum class OutgoingStatusType(val type: Long) {
    // Versioning: Each type supports up to 100 versions.
    // For example, Succeeded_v1(type = 101)
    //
    Succeeded_v0(100),
    Failed_v0(200),
    Mined_v0(300),
    Mined_v1(301); // <= we just add a new version

    fun version(): Long = type % 100L
}
```

If there are existing `Mined_v0` blobs in the database, we can upgrade them to the new structure on-the-fly. In particular, we will create a fake `Part.ClosingTransaction` with values from the old `OnChain` JSON.

Reply:
Good, let's wait to do that later then!

Do note though that with the current code, we may end up showing users a negative fee in the following scenario:

- The channel has a big pending outgoing htlc (to be extreme, all the funds that were on the user side)
- We manage to get the htlc funds back via an htlc-timeout tx on-chain
- `recipientAmount` was `0` because it didn't take this htlc into account
- `claimed` will be `N sats` (the htlc amount)

To correctly address that we'll need to do what I mentioned in another comment: when going to `Closed`, we should check the final state, update `recipientAmount` and calculate accurate fees (but let's do it later).